src/AtlasAdmin.ts

```diff
@@ -0,0 +1,173 @@
+  defaultHeaders: RequestOptions;
+  baseUrl: string;
+
+  constructor(publicApiKey: string, privApiKey: string) {
+    // set base headers
+    this.defaultHeaders = DEFAULT_ATLAS_API_OPTIONS;
+    this.defaultHeaders['digestAuth'] = `${publicApiKey}:${privApiKey}`;
+    // set base url
+    this.baseUrl = `https://cloud.mongodb.com/api/atlas/v1.0/groups/${GROUP_ID}/clusters/${CLUSTER_NAME}/fts/indexes`;
+  }
+
+  async patchSearchIndex(taxonomy: Taxonomy) {
+    log.info('patchSearchIndex');
+    try {
+      const index = await this.findSearchIndex(DB, COLLECTION_NAME, SEARCH_INDEX);
+      if (!index) {
+        return this.createSearchIndex(SearchIndex, taxonomy);
+      }
+      return this.updateSearchindex(index['indexID'], SearchIndex, taxonomy);
+    } catch (e) {
+      log.error(`Error while patching searching index: ${JSON.stringify(e)}`);
+    }
+  }
+
+  private async findSearchIndex(db: string, collection: string, indexName: string) {
+    log.info('finding Atlas search index');
+    const url = `${this.baseUrl}/${db}/${collection}`;
+    const options = DEFAULT_ATLAS_API_OPTIONS;
+    options['method'] = 'GET';
+
+    try {
+      const { data, res } = await request(url, options);
+      if (res.statusCode !== 200) {
+        log.error(res);
+        throw res;
+      }
+
+      return data.find((index: any) => index.name === indexName);
+    } catch (e) {
+      log.error(`Error while fetching search index: ${JSON.stringify(e)}`);
+      throw e;
+    }
+  }
+
+  private async createSearchIndex(searchIndex: IndexMappings, taxonomy: Taxonomy) {
+    log.info('creating Atlas search index');
+    const url = `${this.baseUrl}`;
+    const options = this._getPostPatchOptions(searchIndex, taxonomy);
+    options['method'] = 'POST';
+
+    try {
+      const { res, data } = await request(url, options);
+      if (res.statusCode !== 200) {
+        log.error(res);
+        throw res;
+      }
+
+      return data;
+    } catch (e) {
+      log.error(`Error while creating new search index: ${JSON.stringify(e)}`);
+      throw e;
+    }
+  }
+
+  private async updateSearchindex(indexId: string, searchIndex: IndexMappings, taxonomy: Taxonomy) {
+    log.info('updating Atlas search index');
+    const url = `${this.baseUrl}/${indexId}`;
+
+    const options = this._getPostPatchOptions(searchIndex, taxonomy);
+    options['method'] = 'PATCH';
+
+    try {
+      const { res, data } = await request(url, options);
+      if (res.statusCode !== 200) {
+        log.error(res);
+        throw res;
+      }
+
+      return data;
+    } catch (e) {
+      log.error(`Error while updating search index: ${JSON.stringify(e)}`);
+      throw e;
+    }
+  }
+
+  private _getPostPatchOptions(searchIndex: IndexMappings, taxonomy: Taxonomy) {
+    const options = DEFAULT_ATLAS_API_OPTIONS;
+    options['data'] = this._insertTaxonomyIntoSearchIndex(searchIndex, taxonomy);
+    options['data'] = Object.assign(options['data'], {
```

Comment:
Super NIT: Since `COLLECTION_NAME`, `DB`, and `SEARCH_INDEX` are available globally, could the `_insertTaxonomyIntoSearchIndex` add these properties when it returns the object?

Reply:
i'd actually leave this here as this assigns the data for post/patch options. `insertTaxonomyIntoSearchIndex` should literally convert the taxonomy into search index keys

Reply:
Gotcha that makes sense to me!!